src_feedback/ksym.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::HashMap;
+use std::cell::RefCell;
+use std::io::{BufReader, BufRead, Error};
+use std::ffi::CString;
+use std::fs::File;
+use std::default::Default;
+use std::rc::Rc;
+use std::u64;
+
+use crate::SymResolver;
+
+const KALLSYMS: &str = "/proc/kallsyms";
+const DFL_KSYM_CAP: usize = 200000;
+
+pub struct Ksym {
+    pub addr: u64,
+    pub name: String,
+    pub c_name: RefCell<Option<CString>>,
+}
+
+/// The symbol resolver for /proc/kallsyms.
+///
+/// The users should provide the path of kallsyms, so you can provide
+/// a copy from other devices.
+pub struct KSymResolver {
+    syms: Vec<Ksym>,
+    sym_to_addr: RefCell<HashMap<&'static str, u64>>,
+}
+
+impl KSymResolver {
+    pub fn new() -> KSymResolver {
+	Default::default()
+    }
+
+    pub fn load_file_name(&mut self, filename: &str) -> Result<(), std::io::Error> {
+	let f = File::open(filename)?;
+	let mut reader = BufReader::new(f);
+	let mut line = String::new();
+
+	while let Ok(sz) = reader.read_line(&mut line) {
+	    if sz == 0 {
+		break;
+	    }
+	    let tokens: Vec<&str> = line.split_whitespace().collect();
+	    if tokens.len() < 3 {
+		break;
+	    }
+	    let (addr, _symbol, func) = (tokens[0], tokens[1], tokens[2]);
+	    if let Ok(addr) = u64::from_str_radix(addr, 16) {
+		let name = String::from(func);
+		self.syms.push(Ksym { addr, name, c_name: RefCell::new(None) });
+	    }
+
+	    line.truncate(0);
```

Comment:
btw, I think line.clear() also works; it looks like it calls .truncate(0) underneath (https://doc.rust-lang.org/stable/src/alloc/vec/mod.rs.html#1881)

Reply:
I am not super familiar with Rust.  So, I didn't aware String::clear() is there until later.  Yes, you are right.